src/logic/safe/store/actions/processTransaction.ts

```diff
@@ -63,143 +40,57 @@ interface ProcessTransactionArgs {
-        const signature = await tryOffChainSigning(
-          tx.safeTxHash,
-          { ...txArgs, safeAddress },
-          hardwareWallet,
-          safeVersion,
-        )
-
-        if (signature) {
-          dispatch(closeSnackbarAction({ key: beforeExecutionKey }))
-
-          await saveTxToHistory({ ...txArgs, signature })
-
-          dispatch(fetchTransactions(chainId, safeAddress))
-          return
-        }
+class TxProcessor extends TxSender {
+  public processTransaction(props: ProcessTransactionArgs): ProcessTransactionAction {
+    return async (dispatch: Dispatch, getState: () => AppReduxState): Promise<void> => {
+      // Selectors
+      const state = getState()
+
+      const { tx } = props
+
+      const txProps = {
+        navigateToTransactionsTab: false,
+        notifiedTransaction: props.notifiedTransaction,
+        operation: tx.operation,
+        origin: tx.origin,
+        safeAddress: props.safeAddress,
+        to: tx.to,
+        txData: tx.data ?? EMPTY_DATA,
+        txNonce: '',
+        valueInWei: tx.value,
+        safeTxGas: tx.safeTxGas,
+        ethParameters: props.ethParameters,
+        delayExecution: !props.approveAndExecute,
       }
 
-      transaction = isExecution ? getExecutionTransaction(txArgs) : getApprovalTransaction(safeInstance, tx.safeTxHash)
-
-      const sendParams: PayableTx = {
-        from,
-        value: 0,
-        gas: ethParameters?.ethGasLimit,
-        [getGasParam()]: ethParameters?.ethGasPriceInGWei,
-        nonce: ethParameters?.ethNonce,
+      // Populate instance vars
+      try {
+        await this.prepare(dispatch, state, txProps)
+      } catch (err) {
+        return
       }
 
-      await transaction
-        .send(sendParams)
-        .once('transactionHash', async (hash: string) => {
-          txHash = hash
-          dispatch(closeSnackbarAction({ key: beforeExecutionKey }))
-
-          if (isExecution) {
-            dispatch(updateTransactionStatus({ safeTxHash: tx.safeTxHash, status: LocalTransactionStatus.PENDING }))
-          }
-
-          try {
-            await saveTxToHistory({ ...txArgs })
-
-            // store the pending transaction's nonce
-            isExecution && aboutToExecuteTx.setNonce(txArgs.nonce)
-
-            dispatch(fetchTransactions(chainId, safeAddress))
-          } catch (e) {
-            logError(Errors._804, e.message)
-          }
-        })
-        .then(async (receipt) => {
-          dispatch(fetchTransactions(chainId, safeAddress))
-
-          if (isExecution) {
-            dispatch(fetchSafe(safeAddress))
-          }
-
-          return receipt.transactionHash
-        })
-    } catch (err) {
-      logError(Errors._804, err.message)
-
-      dispatch(closeSnackbarAction({ key: beforeExecutionKey }))
-
-      if (isExecution) {
-        dispatch(updateTransactionStatus({ safeTxHash: tx.safeTxHash, status: LocalTransactionStatus.PENDING_FAILED }))
+      this.safeTxHash = tx.safeTxHash
```

Comment:
Why does it have to be a property on the instance of the processor? it doesn't make sense from OOP point of view

Reply:
Adding info I discovered when talking with Aaron:
so because the same instance is reused across different transactions I think it's possible that methods that are called later may reference the wrong values. Like `this.isExecution` in `onComplete`. If I create a transaction that is supposed to be executed and `isExecution` is set to true, and after that, I create another one where `this.isExecution` is false. When `onComplete` is called for the first transaction `this.isExecution` will be false

Reply:
Haven't test it though, but since the interface allows such scenario it could be good to add a test for such cases with subsequent transactions

Reply:
Oh, that's very no bueno. Singleton vs instantiating the class for every tx. 🤦
Great catch, thanks Mikhail!

Reply:
@usame-algan and I have just adjusted it to instantiate the `TxSender` with every call. It would be great if you can take another look at it please @mikheevm @katspaugh